### PR TITLE
[feat]: [CDS-58159]: integrated overrides v2 variables with service step

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48457)

### CODE REVIEW

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 5 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48457)* status: 10 file(s) to review. 0 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/48457#-NWb8aYZ7aexQQjR79Ip:-NWb8aYZ7aexQQjR79Iq:b-3y6afn) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java#L105)):*
> ```Java
>         ? new ArrayList<>()
>         : ngServiceOverrides.getServiceOverrideInfoConfig().getVariables();
>     final Map<String, Object> variables = isOverrideV2Enabled && isNotEmpty(overridesV2Configs)
> ```

if overrideV2Enabled is true but overrideConfigs are empty, should we go to the else branch ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 1]
```diff
@@ -480,33 +506,47 @@ private void executeEnvironmentPart(Ambiance ambiance, ServiceStepV3Parameters p
...
-          && ngEnvironmentConfig.getNgEnvironmentInfoConfig().getVariables() != null) {
-        secretNGVariables.addAll(ngEnvironmentConfig.getNgEnvironmentInfoConfig()
-                                     .getVariables()
-                                     .stream()
-                                     .filter(SecretNGVariable.class ::isInstance)
-                                     .collect(Collectors.toList()));
-      }
+      if (isOverridesV2enabled) {
+        if (ngEnvironmentConfig != null && ngEnvironmentConfig.getNgEnvironmentInfoConfig() != null) {
+          mergedOverrideV2Configs.put(
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:522]: should it be putIfAbsent() and what happens if the value corresponding ENV_GLOBAL_OVERRIDE was null ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 2]
```diff
@@ -480,33 +506,47 @@ private void executeEnvironmentPart(Ambiance ambiance, ServiceStepV3Parameters p
...
+              ServiceOverridesType.ENV_GLOBAL_OVERRIDE, toOverrideConfigV2(ngEnvironmentConfig, accountId));
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:523]: same as above

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 3]
```diff
@@ -550,6 +591,19 @@ private void executeEnvironmentPart(Ambiance ambiance, ServiceStepV3Parameters p
     }
   }
 
+  @NonNull
+  private static List<NGVariable> getSecretVariablesFromOverridesV2(
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:595]: this need not be static

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 4]
```diff
@@ -571,11 +625,15 @@ private NGServiceOverrideConfig mergeSvcOverrideInputs(
...
 
   private void processServiceVariables(Ambiance ambiance, ServicePartResponse servicePartResponse,
-      NGLogCallback logCallback, EnvironmentOutcome environmentOutcome) {
+      NGLogCallback logCallback, EnvironmentOutcome environmentOutcome, boolean isOverridesV2Enabled,
+      Map<ServiceOverridesType, NGServiceOverrideConfigV2> overridesV2Configs) {
     VariablesSweepingOutput variablesSweepingOutput;
-    if (environmentOutcome != null) {
-      variablesSweepingOutput = getVariablesSweepingOutput(
-          servicePartResponse.getNgServiceConfig().getNgServiceV2InfoConfig(), logCallback, environmentOutcome);
+    if (environmentOutcome != null || (isOverridesV2Enabled && isNotEmpty(overridesV2Configs))) {
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:631]: isn't environmentOutcome not always null ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 5]
```diff
@@ -938,6 +1041,60 @@ protected StepResponse handleFreezeResponse(
     return null;
   }
 
+  private NGServiceOverrideConfigV2 toOverrideConfigV2(
+      NGServiceOverrideConfig configV1, String accountId, NGEnvironmentConfig ngEnvironmentConfig) {
+    NGServiceOverrideInfoConfig serviceOverrideInfoConfig = configV1.getServiceOverrideInfoConfig();
+    NGEnvironmentInfoConfig ngEnvironmentInfoConfig = ngEnvironmentConfig.getNgEnvironmentInfoConfig();
+    final String envRef =
+        IdentifierRefHelper.getRefFromIdentifierOrRef(accountId, ngEnvironmentInfoConfig.getProjectIdentifier(),
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:1049]: nit: order of arguments is not correct

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Comment by tchaurasiya** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209246806:-NWf2LHt0QS9KC0Zlbph:b-ufvhfn) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java#L105)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

if overrideV2Enabled is true but overrideConfigs are empty, should we go to the else branch ?
</blockquote></details>

Done. made changes.

**Comment by tchaurasiya** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java` line 135 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209248298:-NWf5lpe9ojQyBEUnDWm:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentMapper.java#L135)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we should handle other types of overrides
</blockquote></details>

Done.

**Review by tchaurasiya — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48457)* status: 14 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 522 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209254926:-NWf5n460t9-e3zejZn6:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L522)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should it be putIfAbsent() and what happens if the value corresponding ENV_GLOBAL_OVERRIDE was null ?
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 523 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209255224:-NWf5ncM6DMNBOW0xoAe:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L523)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 595 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209255842:-NWf5o0-7o7Di79lu7cp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L595)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this need not be static
</blockquote></details>

Done.
___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 631 at r3](https://reviewable.io/reviews/harness/harness-core/48457#gh-1209258558:-NWf5oQ3DNyH-6Na0arR:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d34347d85e215b5de308a62fb6fe24f3df0bebd4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L631)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

isn't environmentOutcome not always null ?
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 9 files at r1, 3 of 7 files at r4, 3 of 5 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48457)* status: 14 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger uts

**Comment by tchaurasiya** [comment]
trigger ut3

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java [hunk 6]
```diff
@@ -102,6 +106,34 @@ public void prepareAndSaveFinalManifestMetadataToSweepingOutput(@NonNull NGServi
...
+      EnumMap<ServiceOverridesType, NGServiceOverrideConfigV2> overrideV2Configs, String environmentRef) {
+    if (ngServiceV2InfoConfig == null) {
+      throw new InvalidRequestException(SERVICE_CONFIGURATION_NOT_FOUND);
+    }
+    final List<ManifestConfigWrapper> svcManifests = getSvcManifests(ngServiceV2InfoConfig);
+    Map<ServiceOverridesType, List<ManifestConfigWrapper>> manifestsFromOverride = new HashMap<>();
+    if (isNotEmpty(overrideV2Configs)) {
+      overrideV2Configs.forEach((type, override) -> {
+        if (isNotEmpty(override.getSpec().getManifests())) {
+          manifestsFromOverride.put(type, emptyIfNull(override.getSpec().getManifests()));
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java:121]: we have already checked for null in the IF statement

**Review by yogesh-chauhan — APPROVED** [review]

**Review by sourabh-awashti — APPROVED** [review]
lgtm

**Comment by tchaurasiya** [comment]
trigger review

**Pull request merged**